doc/gen_key.1
.TH GEN_KEY 1 "2010 Feb 2" "ASN.1 Tools" "ROA-PKI"

.SH NAME
gen_key \- generates an RSA key pair of specified size

.SH SYNOPSIS
.B gen_key
.I filename size

.SH DESCRIPTION
Generates an RSA key pair with a public modulus of \fIsize\fP bits.
Results are written to to \fIfilename\fP as a cryptlib keyfile (".p15")
with label "label" and password "password".

.SS "FILE FORMAT"
The output keyfile is formatted according to PKCS #15, the
Cryptographic Token Information Format Standard.

.SS "Response Messages"
.nf
    Usage: filename keysize in bits
    Making ... with key size ... bits
    Invalid key size
.fi
where the ellipsis is replaced with specific information.

.SH AUTHOR
Charlie Gardiner.

.SH "SEE ALSO"
.BR add_key_info (1),
.BR sign_cert (1)


ChangeLog
0.12, not yet released

	* Miscellaneous code clean-up and minor bug fixes.
	* Adjust README to instruct users to build a shared version of
	  OpenSSL.
	* Bump copyright year to 2016.
	* Fix unsafe use of snprintf().
	* A compiler with C99 support is now required.
	* Build system improvements.  Autoconf 2.60 is now required.
	* Test suite improvements.

0.11, released 2015-06-12

	* Fix a bug where ROAs with hundreds of ROAIPAddresses or more would be
	  truncated.
	* Add support for new-style Trust Anchor Locators (TALs)
	  [draft-ietf-sidr-rfc6490-bis-03].
	* Fix a potential buffer overflow in code that handles untrusted input.
	* Fix performance scaling of rpki-rtr-daemon. The rpki-rtr-daemon
	  previously had poor performance with very large rtr_incremental or
	  rtr_full tables.
	* Decrease the likelihood of a malformed rsync URI resulting in a
	  successful attack, by passing the --protect-args option to rsync.
	* Improve conformance to [RFC6810] by sending a No Data Available PDU
	  instead of a Cache Reset PDU in response to a serial query for which
	  we don't have data yet.
	* Fix expired certificates in the test suite.
	* Fix a potential infinite loop in one of the self tests.
	* Fix a compatibility bug in the test suite, where the tests were
	  failing on several popular systems that use a particular version of
	  netcat.

0.10, released 2014-02-25

	* Add beta support for running rpstir in 64-bit mode.
	* Add beta support for running rpstir on Mac OS X.
	* Significantly improve performance of incremental updates by adding an
	  rsync flag to preserve file modification times.
	* Fix multiple potential SQL-injection bugs.
	* Support newer versions of rsync and automake.
	* Remove support for RPSL due to lack of demand.
	* Update conformance cases (see doc/conformance-cases).
	* The ./configure script no longer prints a misleading error message
	  when it's run outside of a git directory.
	* Add tests with AS numbers that don't fit into 16 bits.
	* Reject trust anchor CA certificates that have resources marked
	  inherit, as specified in RFC 6490, Section 2.2.
	* Verify that rpstir compiles with clang, and fix all of clang's
	  warnings.
	* In the statistics collection mode, add support for incremental updates
	  and for running multiple simultaneous statistics collections.
	* Change handling of CRLs that list syntactically invalid serial
	  numbers. These CRLs are now accepted, but the certificates with
	  syntactically invalid serial numbers are still not accepted.

0.9, released 2013-10-31

	* Add many more conformance test cases. (See doc/conformance-cases for
	  the full list.) These test cases can be used by relying party
	  software to test compliance with published RFCs and Internet-Drafts.
	  Additionally, relying party software that passes the tests can be
	  used to test the output of a Certification Authority.
	* Fix bugs found by the above test cases.
	* Add a pseudo-random factor to the calculation of how long to wait
	  before retrying an rsync connection. This should help prevent many
	  relying parties from hitting the same server at the same time.
	* Implement basic support for collecting statistics of the RPKI over
	  time.

0.8, released 2013-07-25

	* Support Ghostbusters Records [RFC 6493].
	* Fix a SQL injection bug when a filename in a manifest contains certain
	  characters.
	* Fix a bug where CRLs were not being marked as invalid when their
	  issuers became invalid and manifests were not being marked as invalid
	  when their associated EE certificates became invalid.
	* Fix the query client (rpstir-query) to not consider it an error
	  when no results are returned for a query.
	* Require ROA EE certificates to explicitly specify IP resources.
	  That is, disallow RFC 3779 "inherit" bit in ROA EE certificates.
	* Add functionality to rpstir-results for outputting simple lists
	  of good and bad URIs, to simplify comparison testing.
	* On installation, create ${bindir} if it does not already exist.
	* Improve README based on user feedback.

0.7, released 2013-04-10

	* Add preliminary support for adding files from the local cache. In the
	  future, this could be used to quickly deploy additional relying party
	  machines without each one needing to download all files from the
	  global RPKI.
	* Fix a buffer overflow when manifests list filenames longer than 199
	  bytes.
	* Fix a read of uninitialized memory when the first file listed on a
	  manifest has an unknown extension.
	* Don't allow empty sets of IP resources in certificates.
	  (http://tools.ietf.org/html/rfc6487#section-4.8.10)
	* Don't allow Subsequent Address Family Identifier (SAFI) values in IP
	  resources. (http://tools.ietf.org/html/rfc6487#section-4.8.10)
	* Fix bugs in the ROA validation code where AS numbers equal to zero
	  or larger than 2^31 - 1 were rejected.

0.6, released 2013-02-15

	* Simplify the commands used during normal operation. The commands
	  are now named after what the user intends to do, instead of after the
	  internal component we used to implement them. Some of the scripts are
	  reorganized so that each script does one function (e.g., synchronize
	  with remote repositories) instead of presenting an interface to one
	  internal component (e.g., the component that validates objects fed to
	  it by other components while syncronizing with remote repositories).
	* Use a single configuration file syntax instead of a mix of
	  environment variables and multiple incompatible configuration files.
	  Additionally, provide an example configuration file with comments
	  describing each option. This should make rpstir easier to set up.
	* Switch from one Makefile per directory to a single Makefile. Aside
	  from speeding up parallel compilation, this lets us simplify
	  internal dependencies. With one Makefile per directory, we had to
	  carefully order the directories such that nothing in one directory

cg/tools/gen_key.c
#include <unistd.h>
#include <stdio.h>
#include "cryptlib.h"
#include <string.h>

/*
 * $Id: genkey.c 506 2008-06-03 21:20:05Z gardiner $ 
 */

#define STANDARD_KEY_SIZE 2048


int main(
    int argc,
    char **argv)
{
    CRYPT_CONTEXT privKeyContext;
    CRYPT_KEYSET cryptKeyset;
    int ksize;
    if (argc < 3)
    {
        fprintf(stderr, "Usage: %s filename keysize\n", argv[0]);
        return 1;
    }
    if (sscanf(argv[2], "%d", &ksize) != 1)
    {
        fprintf(stderr, "Invalid key size\n");
        return 1;
    }
    if (ksize != STANDARD_KEY_SIZE)
    {
        fprintf(stderr, "Warning: key size %d is not the standard size (%d)\n",
                ksize, STANDARD_KEY_SIZE);
    }
    printf("Making %s with key size %d bits \n", argv[1], ksize);
    if (cryptInit() != CRYPT_OK)
    {
        fprintf(stderr, "Can't open Cryptlib\n");
        return 1;
    }
    cryptCreateContext(&privKeyContext, CRYPT_UNUSED, CRYPT_ALGO_RSA);
    cryptSetAttributeString(privKeyContext, CRYPT_CTXINFO_LABEL, "label", 5);
    cryptSetAttribute(privKeyContext, CRYPT_CTXINFO_KEYSIZE, ksize / 8);
    cryptGenerateKey(privKeyContext);
    cryptKeysetOpen(&cryptKeyset, CRYPT_UNUSED, CRYPT_KEYSET_FILE,
                    argv[1], CRYPT_KEYOPT_CREATE);
    cryptAddPrivateKey(cryptKeyset, privKeyContext, "password");
    cryptKeysetClose(cryptKeyset);
    cryptDestroyContext(privKeyContext);
    cryptEnd();

    return 0;
}
